File: app/src/main/java/com/orlinskas/notebook/ui/viewModel/ConcreteDayViewModel.java
package com.orlinskas.notebook.ui.viewModel;

import android.app.Application;
import android.os.Handler;

import androidx.annotation.NonNull;
import androidx.lifecycle.AndroidViewModel;
import androidx.lifecycle.LiveData;
import androidx.lifecycle.MutableLiveData;

import com.orlinskas.notebook.App;
import com.orlinskas.notebook.Constants;
import com.orlinskas.notebook.CoroutinesFunKt;
import com.orlinskas.notebook.Enums;
import com.orlinskas.notebook.builder.ToastBuilder;
import com.orlinskas.notebook.entity.Notification;
import com.orlinskas.notebook.repository.NotificationRepository;
import com.orlinskas.notebook.value.Day;

import org.jetbrains.annotations.NotNull;

import java.util.List;
import java.util.concurrent.CancellationException;

import kotlin.Unit;
import kotlin.coroutines.Continuation;
import kotlin.coroutines.CoroutineContext;
import kotlinx.coroutines.BuildersKt;
import kotlinx.coroutines.CoroutineScope;
import kotlinx.coroutines.CoroutineStart;
import kotlinx.coroutines.Dispatchers;
import kotlinx.coroutines.Job;

public class ConcreteDayViewModel extends AndroidViewModel {
    private NotificationRepository repository = App.getInstance().getRepository();
    private LiveData<Enum<Enums.RepositoryStatus>> repositoryStatusData = repository.getRepositoryStatusData();
    private LiveData<Enum<Enums.ConnectionStatus>> connectionStatusData = repository.getConnectionStatusData();
    private LiveData<List<Day>> daysData;
    private Job job;
    private CoroutineScope scope = CoroutinesFunKt.getIoScope();
    private Handler handler = new Handler();

    public ConcreteDayViewModel(@NonNull Application application) {
        super(application);
        job = BuildersKt.launch(scope, scope.getCoroutineContext(), CoroutineStart.DEFAULT,
                (scope, continuation) -> {
                    repository.findActual(System.currentTimeMillis(), new Continuation<Unit>() {
                        @NotNull
                        @Override
                        public CoroutineContext getContext() {
                            return scope.getCoroutineContext();
                        }
                        @Override
                        public void resumeWith(@NotNull Object o) {
                            showConnectionStatus();
                        }
                    });
                    return scope.getCoroutineContext();
                });
    }

    public LiveData<List<Day>> getDaysData()  {
        if(daysData == null) {
            try {
                daysData = BuildersKt.runBlocking(scope.getCoroutineContext(),
                        (scope, continuation) -> repository.fastStart(new Continuation<MutableLiveData<List<Day>>>() {
                            @NotNull
                            @Override
                            public CoroutineContext getContext() {
                                return scope.getCoroutineContext();
                            }
                            @Override
                            public void resumeWith(@NotNull Object o) {
                                showConnectionStatus();
                            }
                        }));
            } catch (InterruptedException e) {
                e.printStackTrace();
            }
        }
        return daysData;
    }

    public LiveData<Enum<Enums.RepositoryStatus>> getRepositoryStatusData() {
        return repositoryStatusData;
    }

    public void deleteNotification(Notification notification) {
        job = BuildersKt.launch(scope, scope.getCoroutineContext(), CoroutineStart.DEFAULT,
                (scope, coroutine) -> repository.delete(notification, new Continuation<Unit>() {
                    @NotNull
                    @Override
                    public CoroutineContext getContext() {
                        return Dispatchers.getIO();
                    }
                    @Override
                    public void resumeWith(@NotNull Object o) {
                        showConnectionStatus();
                    }
                }));
    }

    private void showConnectionStatus() {
        Enum<Enums.ConnectionStatus> status = connectionStatusData.getValue();
        if (status == Enums.ConnectionStatus.CONNECTION_DONE) {
            doToast(Constants.REMOTE);
        }
        if (status== Enums.ConnectionStatus.CONNECTION_FAIL) {
            doToast(Constants.LOCAL);
        }
    }

    private void doToast(String message) {
        handler.post(() -> ToastBuilder.doToast(getApplication().getBaseContext(), message));
    }

    @Override
    protected void onCleared() {
        super.onCleared();
        job.cancel(new CancellationException());
    }
}

File: app/src/main/java/com/orlinskas/notebook/mvvm/viewModel/ConcreteDayViewModel.kt
package com.orlinskas.notebook.mvvm.viewModel

import androidx.lifecycle.MutableLiveData
import androidx.lifecycle.ViewModel
import com.orlinskas.notebook.App.Component.app
import com.orlinskas.notebook.Enums
import com.orlinskas.notebook.interactor.DeleteNotificationUseCase
import com.orlinskas.notebook.interactor.FindActualNotificationUseCase
import com.orlinskas.notebook.mvvm.model.Day
import com.orlinskas.notebook.mvvm.model.Notification
import com.orlinskas.notebook.repository.NotificationRepository
import kotlinx.coroutines.CoroutineScope
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.Job
import kotlinx.coroutines.cancel
import javax.inject.Inject

class ConcreteDayViewModel : ViewModel() {
    @Inject lateinit var repository: NotificationRepository
    @Inject lateinit var downloadStatusData: MutableLiveData<Enum<Enums.DownloadStatus>>
    @Inject lateinit var connectionStatusData: MutableLiveData<Enum<Enums.ConnectionStatus>>
    @Inject lateinit var daysData: MutableLiveData<List<Day>>
    @Inject lateinit var deleteNotificationUseCase: DeleteNotificationUseCase
    @Inject lateinit var findActualNotificationUseCase: FindActualNotificationUseCase

    private val job: Job = Job()
    private val scope = CoroutineScope(Dispatchers.IO + job)

    init {
        app.getComponent().inject(this)
    }

    fun deleteNotification(notification: Notification) {
        downloadStatusData.postValue(Enums.DownloadStatus.LOADING)

        deleteNotificationUseCase(notification) {
            handleConnection(it.code)
            updateDaysData()
            downloadStatusData.postValue(Enums.DownloadStatus.READY)
        }
    }

    private fun updateDaysData() {
        findActualNotificationUseCase(params = System.currentTimeMillis()) {
            try {
                daysData.postValue(it.data as List<Day>?)
                handleConnection(it.code)
            } catch (e: Exception) {
                //error
            }
        }
    }

    private fun handleConnection(code: Int) {
        connectionStatusData.postValue(when (code) {
            200 -> Enums.ConnectionStatus.CONNECTION_DONE
            404 -> Enums.ConnectionStatus.CONNECTION_FAIL
            else -> Enums.ConnectionStatus.CONNECTION_FAIL
        })
    }

    override fun onCleared() {
        super.onCleared()
        scope.cancel()
    }
}
